<a target="_blank" href="https://colab.research.google.com/github/LPolyakova/Linear_Algebra_for_CS_students/blob/main/CP/LA_CP_02_Linear_equations.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

# ЛА-Компьютерний практикум-02. Системи лінійних рівнянь

🧩 **2.6. Взаємодія галузей економіки**

1️⃣  У цій задачі за допомогою систем лінійних рівнянь ми будемо досліджувати як впливає зміна потреб у тих чи інших ресурсах на виробництво цих ресурсів.
Припустимо ми маємо дві галузі економіки (наприклад, виробництво сталі ($B_1$), та виробництво автомобілей ($B_2$)). Кожна з цих галузей $B_i$ виробляє продукцію, яку споживають: вона сама ($d_{ii}$), друга з розгляданих галузей ($d_{ij}$), а також інші галузі економіки ($d_{i0}$). Наприклад, за означеними галузями у деякому році спостерігається ситуація, наведена в таблиці нижче. До речі, ці дані наведено щодо економіки США за 1958 рік у млн.дол. зі статті нобелівського лауреата з економіки Василя Леонтьєва: Wassily W. Leontief, The Structure of the U.S. Economy, Scientific
American, volume 212 number 4 (Apr. 1965), p. 2. Ідея аналізу, відомого також як input-output analysis, також належить йому.

In [3]:
df_economy = pd.DataFrame(columns = ['branch', 'B1_consumption', 'B2_consumption', 'other_consumption', 'total'])
df_economy['branch'] = ['B1', 'B2']
df_economy['B1_consumption'] = [5395, 48]
df_economy['B2_consumption'] = [2664, 9030]
df_economy['other_consumption'] = ['', '']
df_economy['total'] = [25448, 30346]
df_economy

branch  B1_consumption  B2_consumption other_consumption  total
0     B1            5395            2664                    25448
1     B2              48            9030                    30346

З цієї таблиці, зокрема можна довідатися, що сталеливарна промисловість споживає власної продукції на суму $d_{11} = 5395$ млн. дол., а автомобілебудування споживає сталі на суму $d_{12} = 2664$ млн. дол. Також, знаючи загальний обсяг виробництва сталі $D_1=25448$, можна  розрахувати скільки сталі споживають інші галузі економіки (зокрема й іде на експорт, або залишається на складах): $d_{10}=D_1-d_{11}-d_{12}$. Зробіть це для обох галузей $B_1$, $B_2$ й запишіть результати у стовпчик 'other_consumption'.   

In [4]:
#write your code here
df_economy['other_consumption'] = df_economy['total'] - df_economy['B1_consumption'] - df_economy['B2_consumption']
df_economy

branch  B1_consumption  B2_consumption  other_consumption  total
0     B1            5395            2664              17389  25448
1     B2              48            9030              21268  30346

Нехай ми маємо прогноз щодо того, як зміниться споживання сталі та авто зовнішніми споживачами ('other\_consumption') на наступний рік:

$d'_{10} =d_{10} +\Delta _{1}$, $d'_{20} =d_{20} +\Delta _{2}$, $\Delta _{1} =200$, $\Delta _{2} =-25$. Тобто споживання сталі зовнішніми споживачами зросте на 200 млн. дол., а споживання авто зменшиться на 25 млн. дол.



Наша мета розрахувати, скільки сталі та авто має вироблятися для того, щоб задовольнити усі п,отреби зважаючи на передбачувані зміни у споживанні. 



Для цього ми запишемо умови, що пов'язують усі наведені та шукані величини.



Споживання цього року (значення всіх величин нам відомі з таблиці) задається системою умов:



${\displaystyle \begin{cases}
D_{1} =d_{11} +d_{12} +d_{10} ;\\
D_{2} =d_{21} +d_{22} +d_{20}
\end{cases}}$



Споживання наступного року:



${\displaystyle \begin{cases}
D'_{1} =d'_{11} +d'_{12} +d'_{10} ;\\
D'_{2} =d'_{21} +d'_{22} +d'_{20}
\end{cases}}$



У другій системі умов нам відомі лише значення величин $d'_{10} ,d'_{20}$. Для того, щоб знайти $D'_{1} ,D'_{2}$ ми будемо додержуватися припущення, що частка споживання галуззю власної продукції та продукції другої з розгляданих галузей залишається незмінною з року в рік, тобто: $\dfrac{d'_{ij}}{D'_{i}} =\dfrac{d_{ij}}{D_{i}}$. Наприклад, якщо цього року автопромисловість споживала приблизно 10\% усієї сталі, то наступного року вона споживатиме такий самий відіоток.



З останньої рівності виразіть величини $d'_{ij}$ та підставте в другу систему. Також підставте значення $d' _{10} ,d'_{20}$. Має залишитися система рівнянь з двома невідомими $D'_{1} ,D'_{2}$.

Запишіть матрицю системи, її праву частину, та розв'яжіть систему, щоб знайти  $D'_1, D'_2$.



✔️ 
## У матричному вигляді:

Маємо систему рівнянь:

$$ A \cdot X = B $$

де:
- \( A \) — матриця коефіцієнтів,
- \( X \) — вектор невідомих,
- \( B \) — вектор правих частин.

### Матриці:

$$
A = \begin{bmatrix}
\frac{d_{10}}{D_1} & 0 \\
0 & \frac{d_{20}}{D_2} 
\end{bmatrix}  
$$

$$
X = \begin{bmatrix} D'_1 \\ D'_2 \end{bmatrix}  
$$

$$
B = \begin{bmatrix} d'_{10} \\ d'_{20} \end{bmatrix}  
$$

### Обернена матриця \( A^{-1} \):

Обернену матрицю \( A^{-1} \) можна знайти за правилом обернення діагональної матриці:

$$
A^{-1} = \begin{bmatrix}
\frac{D_1}{d_{10}} & 0 \\
0 & \frac{D_2}{d_{20}} 
\end{bmatrix}  
$$

### Розв’язок:

Знайдемо \( X \):

$$
X = A^{-1} \cdot B 
$$
`
$$
\begin{bmatrix} D'_1 \\ D'_2 \end{bmatrix} =
\begin{bmatrix}
\frac{D_1}{d_{10}} & 0 \\
0 & \frac{D_2}{d_{20}} 
\end{bmatrix}
\cdot
\begin{bmatrix} d'_{10} \\ d'_{20} \end{bmatrix}  
$$

Виконаємо множення матриць:

$$
D'_1 = \frac{D_1}{d_{10}} \cdot d'_{10} = D_1 \cdot \frac{d'_{10}}{d_{10}}
$$

$$
D'_2 = \frac{D_2}{d_{20}} \cdot d'_{20} = D_2 \cdot \frac{d'_{20}}{d_{20}}
$$

Таким чином, ми отримали значення D'_1 та D'_2.

### Використані значення з першої частини завдання:

$$
D_1 = 25448, \quad D_2 = 30346
$$

$$
d_{11} = 5395, \quad d_{12} = 2664
$$

$$
d_{21} = 48, \quad d_{22} = 9030
$$

$$
d_{10} = 17389, \quad d_{20} = 21268
$$
### Підставимо наші значення:
$$
D'_1 = \frac{25448}{17389} \times 17589 = 25740.691
$$

$$
D'_2 = \frac{30346}{21268} \times 21243 = 30310.329
$$



Зробіть спостереження: чи вірно, що виробництво сталі ($D'_1$) зросло на величину $\Delta_1$, а виробництво авто ($D'_2$) зменшилося на величину $\Delta_2$? Як змінилися ці величини порівняно з попереднім роком? Чи можете Ви висловити припущення, чому так сталося?

✔️
### Ні, вони змінилися не на значення $\Delta_1$ та $\Delta_2$, вони змінилися на інші значення. Припущення: через те, що у нас частка споживання залишилася такою ж, то значення будуть пропорційно змінюватися(згідно з розв'язком)

2️⃣ Узагальніть ідею такого аналізу на довільну кількість окремо розглянутих галузей: $B_1, B_2, \dots , B_n$ та зовнішніх споживачів ('other consumption').
Вважайте, що вам задано вектор загального виробництва (і споживання) $D = (D_1, \dots , D_n)$, матрицю споживання галузями $(d_{ij})$ на цей рік, а також вектор прогнозів змін у споживанні зовнішніми споживачами на наступний рік $\Delta = (\Delta_1, \dots, \Delta_n)$, такий що $d'_{i0}=d_{i0}+\Delta_i$. Напишіть функцію, яка розраховує вектор виробництва за галузями на наступний рік: $D' = (D'_1, \dots , D'_n)$. 

In [4]:
import numpy as np
import pandas as pd

def calculate_production_next_year(D, consumption_matrix, delta):
    """
    Розраховує вектор виробництва на наступний рік для n галузей.
    
    Parameters:
        D (list or np.array): Вектор загального виробництва (D_1, ..., D_n) цього року.
        consumption_matrix (np.array): Матриця споживання (d_ij), розміром n x n.
        delta (list or np.array): Вектор прогнозованих змін зовнішнього споживання (Δ_1, ..., Δ_n).
    
    Returns:
        np.array: Вектор виробництва на наступний рік (D'_1, ..., D'_n).
    """
    n = len(D)
    D = np.array(D)
    delta = np.array(delta)
    consumption_matrix = np.array(consumption_matrix)
    
    # Обчислюємо зовнішнє споживання цього року
    external_consumption = D - np.sum(consumption_matrix, axis=1)
    
    # Прогнозоване зовнішнє споживання наступного року
    external_consumption_next_year = external_consumption + delta
    
    # Обчислюємо коефіцієнти a_i
    coefficients = 1 - np.sum(consumption_matrix / D[:, None], axis=1)
    
    # Обчислюємо виробництво на наступний рік
    D_next_year = external_consumption_next_year / coefficients
    
    return D_next_year

# Тестовий приклад
if __name__ == "__main__":
    # Вхідні дані з таблиці
    D = [800, 300, 500]  # Вектор загального виробництва цього року
    consumption_matrix = [
        [25, 50, 100],  # Споживання B1
        [25, 50, 50],   # Споживання B2
        [15, 10, 0]     # Споживання B3
    ]
    delta = [625, 200, 475]  # Вектор прогнозованих змін зовнішнього споживання

    # Виклик функції
    D_next_year = calculate_production_next_year(D, consumption_matrix, delta)

    # Формуємо таблицю для відображення результатів
    df_result = pd.DataFrame({
        'Branch': ['B1', 'B2', 'B3'],
        'Current Production (D)': D,
        'Consumption (B1)': consumption_matrix[0],
        'Consumption (B2)': consumption_matrix[1],
        'Consumption (B3)': consumption_matrix[2],
        'Delta': delta,
        'New Production (D\')': D_next_year
    })

    # Виводимо таблицю
    print("Таблиця результатів:")
    print(df_result)


Таблиця результатів:
  Branch  Current Production (D)  Consumption (B1)  Consumption (B2)  \
0     B1                     800                25                25   
1     B2                     300                50                50   
2     B3                     500               100                50   

   Consumption (B3)  Delta  New Production (D')  
0                15    625          1600.000000  
1                10    200           642.857143  
2                 0    475          1000.000000  


Де 1600 - споживання B1,
642.85714286 - споживання B2,
1000 - - споживання B3.

Протестуйте вашу функцію зокрема на наступній матриці статистики виробництва-споживання за трьома галузями (наприклад, 'сільське господарство', 'залізниця', 'морські перевезення') і векторі змін зовнішнього споживання $\Delta = (625, 200, 475)$

In [5]:
df_economy_test = pd.DataFrame(columns = ['branch', 'B1_consumption', 'B2_consumption','B3_consumption' ,'other_consumption', 'total'])
df_economy_test['branch'] = ['B1', 'B2', 'B3']
df_economy_test['B1_consumption'] = [25,  25, 15]
df_economy_test['B2_consumption'] = [50,  50, 10]
df_economy_test['B3_consumption'] = [100, 50, 0]
df_economy_test['other_consumption'] = ['', '', '']
df_economy_test['total'] = [800, 300, 500]
df_economy_test

branch  B1_consumption  B2_consumption  B3_consumption other_consumption  \
0     B1              25              50             100                     
1     B2              25              50              50                     
2     B3              15              10               0                     

   total  
0    800  
1    300  
2    500

### Використані джерела - 
https://web.posibnyky.vntu.edu.ua/fksa/14moskvina__komp_metod_dosl_analiz_danykh/lek2.htm

https://ukrayinska.libretexts.org/%D0%9C%D0%B0%D1%82%D0%B5%D0%BC%D0%B0%D1%82%D0%B8%D0%BA%D0%B0/%D0%9F%D1%80%D0%B8%D0%BA%D0%BB%D0%B0%D0%B4%D0%BD%D0%B0_%D0%BC%D0%B0%D1%82%D0%B5%D0%BC%D0%B0%D1%82%D0%B8%D0%BA%D0%B0/%D0%9F%D1%80%D0%B8%D0%BA%D0%BB%D0%B0%D0%B4%D0%BD%D0%B0_%D1%81%D0%BA%D1%96%D0%BD%D1%87%D0%B5%D0%BD%D0%BD%D0%B0_%D0%BC%D0%B0%D1%82%D0%B5%D0%BC%D0%B0%D1%82%D0%B8%D0%BA%D0%B0_(Sekhon_%D1%96_Bloom)/02%3A_%D0%9C%D0%B0%D1%82%D1%80%D0%B8%D1%86%D1%96/2.02%3A_%D0%A1%D0%B8%D1%81%D1%82%D0%B5%D0%BC%D0%B8_%D0%BB%D1%96%D0%BD%D1%96%D0%B9%D0%BD%D0%B8%D1%85_%D1%80%D1%96%D0%B2%D0%BD%D1%8F%D0%BD%D1%8C_%D1%82%D0%B0_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4_%D0%93%D0%B0%D1%83%D1%81%D0%B0-%D0%99%D0%BE%D1%80%D0%B4%D0%B0%D0%BD%D0%B0


https://jasonwarta.github.io/latex-matrix/ для генерування матриць у LaTEX форматі


ChatGpt для розмітки в markdown та для допомоги з кодом на python